src/sentry/taskworker/taskworker.py

```diff
@@ -0,0 +1,190 @@
+        self, activation: TaskActivation, processing_deadline: datetime
+    ) -> tuple[TaskActivation | None, datetime | None]:
+        from sentry.taskworker.service.client import task_client
+
+        if not self.namespace.contains(activation.taskname):
+            logger.exception("Could not resolve task with name %s", activation.taskname)
+            return (None, None)
+
+        # TODO: Check idempotency
+        task_added_time = activation.received_at.seconds
+        next_state = TASK_ACTIVATION_STATUS_FAILURE
+        result = None
+        try:
+            task_data_parameters = orjson.loads(activation.parameters)
+            processing_timeout = (processing_deadline - datetime.now()).total_seconds()
+            result = self._pool.apply_async(
+                func=taskworker_process._process_activation,
+                args=(
+                    self.options["namespace"],
+                    activation.taskname,
+                    task_data_parameters["args"],
+                    task_data_parameters["kwargs"],
+                ),
+            )
+            # Will trigger a TimeoutError if the task execution runs long
+            result.get(timeout=processing_timeout)
+            taskregistry.get(self.options["namespace"]).get(activation.taskname)(
+                *task_data_parameters["args"], **task_data_parameters["kwargs"]
+            )
+
+            next_state = TASK_ACTIVATION_STATUS_COMPLETE
+        except TimeoutError:
+            logger.info("taskworker.task_execution_timeout")
+            # When a task times out we kill the entire pool. This is necessary because
+            # stdlib multiprocessing.Pool doesn't expose ways to terminate individual tasks.
+            self._pool.terminate()
+            self._build_pool()
+        except Exception as err:
+            logger.info(
+                "taskworker.task_errored", extra={"type": str(err.__class__), "error": str(err)}
+            )
+            # TODO check retry policy
+            if self.namespace.get(activation.taskname).should_retry(activation.retry_state, err):
+                logger.info("taskworker.task.retry", extra={"task": activation.taskname})
+                next_state = TASK_ACTIVATION_STATUS_RETRY
+
+        execution_time = time.time()
+        task_latency = execution_time - task_added_time
+        self._execution_count += 1
+
+        if next_state == TASK_ACTIVATION_STATUS_COMPLETE:
+            logger.info(
+                "taskworker.task.complete", extra={"task": activation.taskname, "id": activation.id}
+            )
+            response = task_client.complete_task(task_id=activation.id)
+        else:
+            logger.info(
+                "taskworker.task.change_status",
+                extra={"task": activation.taskname, "state": next_state},
+            )
+            response = task_client.set_task_status(
+                task_id=activation.id,
+                task_status=next_state,
+            )
```

Comment:
Why do we get another task back again immediately? I remember us talking about this but I don't quite remember why. Looking at the actual code, it feels like this is a little too hard to understand and maintain.

Reply:
bit of a nitpick, but I think if we want to do both in 1 grpc call, could we separate the execution of the task and sending result back to the server into 2 functions?

Reply:
Getting the 'next' task back in the response of a status change lets the worker only make a single RPC call to advance, instead of needing to make one call for status update and another to get the next task.